src/network/networkd-state-file.c

```diff
@@ -874,6 +874,17 @@ static int link_save(Link *link) {
 
                 /************************************************************/
 
+                if (link->dhcp_lease) {
+                        r = sd_dhcp_lease_get_hostname(link->dhcp_lease, &hostname);
+                        if (r < 0 && r != -ENODATA)
+                                return r;
+                }
+
+                if (hostname)
+                        fprintf(f, "HOSTNAME=%s\n", hostname);
```

Comment:
Uh, state file will be deprecated (soon?). Please do not add new entry here.
Please add the hostname entry in json output.
Please update `dhcp_client_append_json()` in networkd-json.c